app/src/main/java/org/wikipedia/gallery/GalleryActivity.java

```diff
@@ -542,11 +557,60 @@ public void layOutGalleryDescription() {
         }
         galleryAdapter.notifyFragments(galleryPager.getCurrentItem());
 
+        if (imageCaptionDisposable != null && !imageCaptionDisposable.isDisposed()) {
+            imageCaptionDisposable.dispose();
+        }
+
+        // TODO: replace when the Media endpoint gives us structured captions automatically.
+        // TODO: remove feature flag when ready.
+
+        if (ReleaseUtil.isPreBetaRelease()) {
+            imageCaptionDisposable = MediaHelper.INSTANCE.getImageCaptions(item.getTitles().getCanonical())
+                    .subscribeOn(Schedulers.io())
+                    .observeOn(AndroidSchedulers.mainThread())
+                    .subscribe(this::updateGalleryDescription, t -> {
+                        L.e(t);
+                        updateGalleryDescription(null);
+                    });
+        } else {
+            updateGalleryDescription(null);
+        }
+    }
+
+    public void updateGalleryDescription(@Nullable Map<String, String> captions) {
+        GalleryItem item = getCurrentItem();
+        if (item == null) {
+            infoContainer.setVisibility(View.GONE);
+            return;
+        }
+        galleryAdapter.notifyFragments(galleryPager.getCurrentItem());
+
         CharSequence descriptionStr = "";
-        if (item.getDescription() != null && item.getDescription().getHtml() != null) {
-            descriptionStr = StringUtil.fromHtml(item.getDescription().getHtml());
-        } else if (item.getDescription() != null && item.getDescription().getText() != null) {
-            descriptionStr = item.getDescription().getText();
+
+        // If we have a structured caption in our current language, then display that instead
+        // of the unstructured description, and make it editable.
+        if (captions != null && captions.containsKey(app.getAppOrSystemLanguageCode())) {
+            descriptionStr = captions.get(app.getAppOrSystemLanguageCode());
+            captionEditButton.setVisibility(View.VISIBLE);
+
+            // and if we have another language in which the caption doesn't exist, then offer
+            // it to be translatable.
+            boolean allowTranslate = false;
+            for (String lang : app.language().getAppLanguageCodes()) {
+                if (!captions.containsKey(lang)) {
+                    allowTranslate = true;
```

Comment:
Should we add a break if it is true?

Reply:
Or, maybe we can set `allowTranslate = true` as an initial value, and then add the `break` in the `if (captions.containsKey(lang))` logic.

Reply:
Sure!